json.dumps fails for numpy int64, if used in python 3.6, aiida 1.0.0b2

when json.dumps is called on a int64 numpy value, an error is raised.

The use that I encountered is raised as 

 File "/aiida_core/aiida/backends/djsite/db/models.py", line 820, in create_value jsondata = json.dumps(value)
  File "/aiida_core/aiida/common/json.py", line 47, in dumps return simplejson.dumps(data, ensure_ascii=False, encoding='utf8', **kwargs)
  File "/aiida_1.0/lib/python3.6/site-packages/simplejson/__init__.py", line 399, in dumps 
    **kw).encode(obj) 
  File "/aiida_1.0/lib/python3.6/site-packages/simplejson/encoder.py", line 296, in encode     chunks = self.iterencode(o, _one_shot=True)
  File "/aiida_1.0/lib/python3.6/site-packages/simplejson/encoder.py", line 378, in iterencode     return _iterencode(o, 0)
  File "/aiida_1.0/lib/python3.6/site-packages/simplejson/encoder.py", line 273, in default     o.__class__.__name__)
TypeError: Object of type int64 is not JSON serializable


A  workaround that worked for me is an additional 'isinstance' check, and conversion to ''int' before dumping (I assume other numpy types will raise the same issue)

I had the very same issue in our AiiDA-KKR plugin, where in the parsers of calculations we regularly use numpy arrays and then got errors like this (in python3 only):
```
E               ValueError: Unable to store the value: it must be either a basic datatype, or json-serializable: -1
```

I currently use this fix on the aiida-core side (in `aiida/backends/djsite/db/models.py`):
```
+++ b/aiida/backends/djsite/db/models.py
@@ -697,10 +697,12 @@ class DbMultipleValueAttributeBaseClass(m.Model):
           bulk_create() call).
         """
         import datetime
+        import numpy as np
 
         from aiida.common import json
         from aiida.common.timezone import is_naive, make_aware, get_current_timezone
 
+
         if cls._subspecifier_field_name is None:
             if subspecifier_value is not None:
                 raise ValueError("You cannot specify a subspecifier value for "
@@ -735,7 +737,7 @@ class DbMultipleValueAttributeBaseClass(m.Model):
             new_entry.fval = None
             new_entry.dval = None
 
-        elif isinstance(value, six.integer_types):
+        elif isinstance(value, six.integer_types) or isinstance(value, np.integer):
             new_entry.datatype = 'int'
             new_entry.ival = value
             new_entry.tval = ''
@@ -743,7 +745,7 @@ class DbMultipleValueAttributeBaseClass(m.Model):
             new_entry.fval = None
             new_entry.dval = None
 
-        elif isinstance(value, float):
+        elif isinstance(value, float) or isinstance(value, np.floating):
             new_entry.datatype = 'float'
             new_entry.fval = value
             new_entry.tval = ''
```

Since aiida-core already has numpy as a dependency why not allow to have numpy datatypes there as well?

--- a/aiida/orm/utils/node.py
+++ b/aiida/orm/utils/node.py
@@ -179,6 +179,9 @@ def clean_value(value):
         if isinstance(val, numbers.Real) and (math.isnan(val) or math.isinf(val)):
             # see https://www.postgresql.org/docs/current/static/datatype-json.html#JSON-TYPE-MAPPING-TABLE
             raise exceptions.ValidationError("nan and inf/-inf can not be serialized to the database")
+        # This fixes #2773 - in python3, ``numpy.int64(-1)`` cannot be json-serialized
+        if isinstance(val, numbers.Integral) and not isinstance(val, bool):
+            val = int(val)
 
         return val
 

